### PR TITLE
[IMPROVEMENT] Refactor SubscriptionsViewController

### LINKED ISSUE #1777: [BUG][NEW UI] Subscription list changing order mistakenly when scrolling
When scrolling up and down the subscription list changes its order. Above my sorting setup and a gif demonstrating the bug (look at the first two subscriptions and how they change as I simply scroll up and down)

screen shot 2018-06-15 at 10 15 07

### PR DESCRIPTION
**Pull request by cardoso**
@RocketChat/ios

Organizing the code better in a MVVM architecture.
Allowing proper animations on removing, inserting and updating rows.

Closes #1777

### CODE REVIEW

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -688,7 +690,7 @@
 		99923773204B3BD800C2D15F /* UploadAvatarRequest.swift in Sources */ = {isa = PBXBuildFile; fileRef = 99923772204B3BD800C2D15F /* UploadAvatarRequest.swift */; };
 		999483EB20644CC4004F61CA /* WebBrowserViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 999483EA20644CC4004F61CA /* WebBrowserViewModel.swift */; };
 		999F25B5207EEB0C002E0F68 /* FileModelMapping.swift in Sources */ = {isa = PBXBuildFile; fileRef = 999F25B4207EEB0C002E0F68 /* FileModelMapping.swift */; };
-		999F25B7207EEE70002E0F68 /* BuildFile in Sources */ = {isa = PBXBuildFile; };
+		999F25B7207EEE70002E0F68 /* (null) in Sources */ = {isa = PBXBuildFile; };
```
Comment (rafaelks) [inline Rocket.Chat.xcodeproj/project.pbxproj]: 👀

File: Rocket.Chat/Controllers/Subscriptions/SubscriptionsList/SubscriptionsViewController.swift [hunk 2]
```diff
@@ -573,33 +323,19 @@ extension SubscriptionsViewController: UITableViewDataSource {
 extension SubscriptionsViewController: UITableViewDelegate {
```
Comment (Sameesunkaria) [inline Rocket.Chat/Controllers/Subscriptions/SubscriptionsList/SubscriptionsViewController.swift]: I believe we can remove this extension all-together.

File: Rocket.Chat.xcodeproj/project.pbxproj (hunk 1 above)
Comment (Sameesunkaria) [reply Rocket.Chat.xcodeproj/project.pbxproj]: I think we should remove that line.

File: Rocket.Chat/Controllers/Subscriptions/SubscriptionsList/SubscriptionsViewController.swift (hunk 2 above)
Comment (cardoso) [reply Rocket.Chat/Controllers/Subscriptions/SubscriptionsList/SubscriptionsViewController.swift]: Why?

Comment (Sameesunkaria) [reply Rocket.Chat/Controllers/Subscriptions/SubscriptionsList/SubscriptionsViewController.swift]: Sorry marked the incorrect extension. It's declaration in not visible on the diff. 😅 
I meant the one starting at line 359.

File: Rocket.Chat/Helpers/RealmAssorter.swift [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+        self.primaryKey = primaryKey
+        self.results = results ?? realm.objects(Object.self)
+        self.sections = []
+        self.token = nil
+
+        self.token = self.results.observe { changes in
+            switch changes {
+            case .update(let values, let deletions, let insertions, let modifications):
+                self.handleUpdate(values: values, deletions: deletions, insertions: insertions, modifications: modifications)
```
Comment (filipealva) [inline Rocket.Chat/Helpers/RealmAssorter.swift]: The logout crash is happening on the first line of `handleUpdate` method, I guess that it's being mistakenly held on memory (at the time of the log out it should be gone, or at least the token invalidated), but it's just a guess. I tried to capture `weak self` but no success.

**Comment by Sameesunkaria** [comment]
@cardoso The sorting header view is not getting themed properly.

File: Rocket.Chat/Helpers/RealmAssorter.swift (hunk 3 above)
Comment (cardoso) [reply Rocket.Chat/Helpers/RealmAssorter.swift]: Fixed :+1: see if you can reproduce again, please

**Comment by cardoso** [comment]
@rafaelks @filipealva @Sameesunkaria fixed the problems you mentioned

Also fixed:
- another possible crash 
- list sometimes beginning empty
- avatars staying black sometimes until cell reload

Seems pretty stable to me now. Please review again :+1:

**Comment by rafaelks** [comment]
@cardoso Thanks for the fixes, but I still have this issues:

**New conversations aren't getting added to the list**


**New messages aren't getting updated on the list**

**Comment by rafaelks** [comment]
@cardoso Just downloaded your changes, and found this issue:

**Pull request merged**